pkg/alignchecker/alignchecker.go

```diff
@@ -14,152 +14,239 @@
+				if n == "" {
+					if t, ok := field.Type.(*dwarf.StructType); ok {
+						if t.Kind == "union" {
+							n = fmt.Sprintf("$union%d", unionCount)
+							unionCount++
+						}
+					}
+				}
+				offsets[n] = field.ByteOffset
+			}
+			structs[st.StructName] = structInfo{
+				size:         st.ByteSize,
+				fieldOffsets: offsets,
+			}
+		}
+	}
+
+	return structs, nil
+}
+
+func check(toCheck map[string]reflect.Type, structs map[string]structInfo) error {
+	for name, g := range toCheck {
+		c, found := structs[name]
+		if !found {
+			return fmt.Errorf("C struct %s not found", name)
+		}
+
+		if c.size != int64(g.Size()) {
+			return fmt.Errorf("struct sizes do not match: %s (%d) vs %s (%d)",
+				g, g.Size(), name, c.size)
+		}
+
+		for i := 0; i < g.NumField(); i++ {
+			fieldName := g.Field(i).Tag.Get("align")
+			// Ignore fields without `align` struct tag
+			if fieldName == "" {
+				continue
+			}
+			goOffset := int64(g.Field(i).Offset)
+			cOffset := structs[name].fieldOffsets[fieldName]
+			if goOffset != cOffset {
+				return fmt.Errorf("%s.%s offset (%d) does not match with %s.%s (%d)",
+					g, g.Field(i).Name, goOffset, name, fieldName, cOffset)
+			}
+		}
+	}
+
+	return nil
 }
 
-func init() {
-	for k, v := range cToGO {
-		err := compareStructs(k, v)
+// Adopted from elf.File.DWARF
+// https://github.com/golang/go/blob/master/src/debug/elf/file.go (go1.11.2).
+//
+// The former method function tries to apply relocations when extracting DWARF
+// debug sections. Unfortunately, the relocations are not implemented for EM_BPF,
+// so the method fails.
+//
+// For struct alignment checks, no relocations are needed, so we comment out
+// the relocation bits.
+//
+// NOTE: DO NOT USE THE FUNCTION FOR ANYTHING ELSE!
+func getDWARFFromELF(f *elf.File) (*dwarf.Data, error) {
+	dwarfSuffix := func(s *elf.Section) string {
+		switch {
+		case strings.HasPrefix(s.Name, ".debug_"):
+			return s.Name[7:]
+		case strings.HasPrefix(s.Name, ".zdebug_"):
+			return s.Name[8:]
+		default:
+			return ""
+		}
+
+	}
+	// sectionData gets the data for s, checks its size, and
+	// applies any applicable relations.
+	sectionData := func(i int, s *elf.Section) ([]byte, error) {
+		b, err := s.Data()
+		if err != nil && uint64(len(b)) < s.Size {
+			return nil, err
+		}
+
+		if len(b) >= 12 && string(b[:4]) == "ZLIB" {
+			dlen := binary.BigEndian.Uint64(b[4:12])
+			dbuf := make([]byte, dlen)
+			r, err := zlib.NewReader(bytes.NewBuffer(b[12:]))
+			if err != nil {
+				return nil, err
+			}
+			if _, err := io.ReadFull(r, dbuf); err != nil {
+				return nil, err
+			}
+			if err := r.Close(); err != nil {
+				return nil, err
+			}
+			b = dbuf
+		}
+
+		for _, r := range f.Sections {
+			if r.Type != elf.SHT_RELA && r.Type != elf.SHT_REL {
+				continue
+			}
+			if int(r.Info) != i {
+				continue
+			}
+			//rd, err := r.Data()
+			_, err := r.Data()
+			if err != nil {
+				return nil, err
+			}
+			// err = f.applyRelocations(b, rd)
```

Comment:
Any particular reason to retain commented-out code?

Reply:
Just wanted to keep it for a reader to show where the relocations were applied in the original code.